ocaml/xapi/xapi_psr.ml

```diff
@@ -329,61 +465,72 @@ end)
 
 module Ptoken = Genptokenlib.Lib
 
-(* at most one PSR should be
-   running at any given time *)
-let m = Mutex.create ()
-
-let start ~__context =
-  (* checked preconditions:
-     * the feature is enabled (to prevent a PSR when not all hosts have upgraded to this version)
-     * we are the master
-     * HA disabled
-     * RPU not running
-     * all hosts in pool 'alive' *)
-  let f () =
-    Pool_features.assert_enabled ~__context ~f:Features.Pool_secret_rotation ;
-    if
-      not
-        (Helpers.is_pool_master ~__context
-           ~host:(Helpers.get_localhost ~__context))
-    then
+let start =
+  let m = Mutex.create () in
+  let with_lock f =
+    (* prevents multiple concurrent PSRs *)
+    if Mutex.try_lock m then (
+      try f () ; Mutex.unlock m with e -> Mutex.unlock m ; raise e
+    ) else
       raise
         Api_errors.(
-          Server_error (host_is_slave, [Pool_role.get_master_address ()])) ;
-    let live_hosts = Helpers.get_live_hosts ~__context |> HostSet.of_list in
-    let all_hosts_list = Xapi_pool_helpers.get_master_slaves_list ~__context in
-    let all_hosts = all_hosts_list |> HostSet.of_list in
-    let offline_hosts = HostSet.diff all_hosts live_hosts in
-    ( if not (HostSet.is_empty offline_hosts) then
-        let offline_host = HostSet.min_elt offline_hosts in
+          Server_error (internal_error, ["pool secret rotation already running"]))
+  in
+  fun ~__context ->
```

Comment:
It's a bit difficult to see in diff here, this construct is used to create a closure with the mutex m and and the function with_lock outside of `start ~__context`, isn't it?

I think that defining them at the top level would be easier to understand, is there a reason why it wasn't done that way?

Reply:
Yes, that's right. Personally I prefer putting it in the scope of `start`, that way it's obvious that only `start` can be using it (i.e. one less thing to consider if we were to run into a locking issue in the future)